src/core/undefined_struct.rs

```diff
@@ -258,6 +258,9 @@ impl From<Vec<u8>> for UndefinedStructTable {
             let mut a: bool;
             let mut b = true;
             loop {
+                if next_index >= len {
```

Comment:
This entire function is a bit hacky as it has been not only ported but converted to a loop rather than a "step to the next item" function.  I have been thinking of refactoring it entirely and a more functional coding style would make it more readable.

We can take this change.  I assume you ran into an out of bounds array condition? I've seen it happen once before, there is a bug in this block but have not been able to spot it.

This block is looking for \0\0 where you've added the len check.  The idea was to not have to keep checking for len by guaranteeing there will be a stop condition.

The stop condition guarantee is lines 242 and 243 (and assumes the \0\0 logic works correctly):
            || data[len - 2] != 0 // 2nd to last byte should be zero and it is not
            || data[len - 1] != 0

Those two lines shouldn't be in a loop (another bug I see I introduced when converting from a single step function to a loop function).

Reply:
I agree. I think you are the best person to know how to improve it.